lib/types.nix

```diff
@@ -107,22 +107,83 @@ rec {
       merge = mergeEqualOption;
     };
 
-    int = mkOptionType rec {
-      name = "int";
-      description = "integer";
-      check = isInt;
-      merge = mergeOneOption;
-    };
+### INTEGERS
+
+    int = ints.signed;
+
+    # specialized subdomains of int
+    ints =
+      let
+        int = mkOptionType rec {
+          name = "int";
+          description = "signed integer";
+          check = isInt;
+          merge = mergeOneOption;
+        };
+
+        betweenDesc = lowest: highest:
+          "${toString lowest} and ${toString highest} (both inclusive)";
+        between = lowest: highest: assert lowest <= highest;
+          addCheck int (x: x >= lowest && x <= highest) // {
+            name = "intBetween";
+            description = "Integer between ${betweenDesc lowest highest}.";
+          };
+        ign = lowest: highest: name: docStart:
+          between lowest highest // {
+            inherit name;
+            description = docStart + " Between ${betweenDesc lowest highest}.";
+          };
+        unsign = bit: range: ign 0 (range - 1)
+          "unsignedInt${toString bit}" "${toString bit} bit unsigned integer.";
+        sign = bit: range: ign (0 - (range / 2)) (range / 2 - 1)
+          "signedInt${toString bit}" "${toString bit} bit signed integer.";
+
+      in rec {
+        /* An int with a fixed range.
+
+          Example:
+            (ints.between 0 100).check (-1)
+            => false
+            (ints.between 0 100).check (101)
+            => false
+            (ints.between 0 0).check 0
+            => true
+        */
+        inherit between;
+
+        unsigned = addCheck types.int (x: x >= 0) // {
+          name = "unsignedInt";
+          description = "unsigned integer, meaning >=0";
+        };
+        unsigned8 = unsign 8 256;
+        unsigned16 = unsign 16 65536;
+        unsigned32 = unsign 32 4294967296;
+        # the biggest int the nix lexer accepts is 9223372036854775808
+        # the smallest int the nix lexer accepts is -9223372036854775807
+        # unsigned64 = unsign 64 18446744073709551616;
+
+        signed = int;
+        signed8 = sign 8 256;
+        signed16 = sign 16 65536;
+        signed32 = sign 32 4294967296;
+
+      };
 
+    /* A network port number. */
+    port = ints.unsigned16;
```

Comment:
https://www.grc.com/port_0.htm

```
"Port Zero" does not officially exist. […]
The designers of the original Berkeley UNIX "Sockets" interface, upon which much of the technology and practice we use today is based, set aside the specification of "port 0" to be used as a sort of "wild card" port. When programming the Sockets interface, the provision of a zero value is generally taken to mean "let the system choose one for me". Programmers who specify "port 0" know that it is an invalid port. They are asking the operating system to pick and assign whatever non-zero port is available and appropriate for their purpose.
```

In favour of simplicity I’d not overcomplicate the option type.
cc @aszlig